docs/reinterpreting_effects.md

```diff
@@ -0,0 +1,359 @@
+A steady diet of dog food may cause blindness in your cat - it lacks taurine.
+Cat owners are 25% likely to pick George Harrison as their favorite Beatle.
+The catnip plant contains an oil called hepetalactone which does for cats what marijuana does to some people. Not all cats react to it those that do appear to enter a trancelike state. A positive reaction takes the form of the cat sniffing the catnip, then licking, biting, chewing it, rub & rolling on it repeatedly, purring, meowing & even leaping in the air.
+The color of the points in Siamese cats is heat related. Cool areas are darker.
+Cats have free-floating clavicle bones that attach their shoulders to their forelimbs, which allows them to squeeze through very small spaces.
+Wikipedia has a recording of a cat meowing, because why not?
+```
+
+### Testing with alternative effect handlers
+
+Per point 2. of our initial implementation criteria, we want to be able to simulate failure cases for testing purposes. This is a great
+case for swapping in an alternative effect handler for our HTTP layer.
+
+This time let's go from the bottom up. In situations where IO is involved, failure scenarios tend
+to surface from least-pure parts of code. In this case, we should therefore implement some facilities
+to experiment with the most failure-prone area: the network itself.
+
+``` haskell
+newtype MockHttpClient m a = MockHttpClient { runMockHttpClient :: ReaderC (HTTP.Request -> IO (HTTP.Response L.ByteString)) m a }
+ deriving newtype
+      ( Monad
+      , Functor
+      , Applicative
+      , MonadIO
+      , Alternative
+      )
+
+runMockHttp :: (HTTP.Request -> IO (HTTP.Response L.ByteString)) -> MockHttpC m a -> m a
+runMockHttp responder m = runReader responder (runMockHttpClient m)
+
+instance (MonadIO m, Algebra sig m) => Algebra (Http :+: sig) (MockHttpClient m) where
+  eff (L (SendRequest req k)) = MockHttpClient ask >>= \responder -> liftIO (responder req) >>= k
+  eff (R other) = MockHttpClient (eff (R (handleCoercible other)))
+
+faultyNetwork :: HTTP.Request -> IO (HTTP.Response L.ByteString)
+faultyNetwork req = throwIO (HTTP.HttpExceptionRequest req HTTP.ConnectionTimeout)
+
+wrongContentType :: HTTP.Request -> IO (HTTP.Response L.ByteString)
+wrongContentType req = pure resp
+  where
+    resp = Response
+      { responseStatus = ok200
+      , responseVersion = http11
+      , responseHeaders = [("Content-Type", "text/xml")]
+      , responseBody = "[{\"text\": \"Cats are not dogs\"}]"
+      , responseCookieJar = mempty
+      , responseClose' = ResponseClose (pure ())
+      }
+
+badJson :: HTTP.Request -> IO (HTTP.Response L.ByteString)
+badJson req = pure resp
+  where
+    resp = Response
+      { responseStatus = ok200
+      , responseVersion = http11
+      , responseHeaders = [("Content-Type", "application/json; charset=utf-8")]
+      , responseBody = "{}"
+      , responseCookieJar = mempty
+      , responseClose' = ResponseClose (pure ())
+      }
```

Comment:
This is _super_ nitpicky, so feel free to ignore, but I don’t feel like the `where` clause is buying us much. We could simplify these to e.g.:

```haskell
badJson req = pure Response
  { responseStatus = ok200
  , …
  }
```

due to the unreasonably high precedence of record syntax.

![Meme of southern preacher exhorting his congregation, captioned “the precedence of record update syntax is TOO DAMN HIGH”](https://pbs.twimg.com/media/CsQVLdiUMAAGYKj?format=jpg&name=small)

Reply:
TIL